### DISCUSSION TITLE
Resizeable nodes [new feature]

### DISCUSSION THREAD

**Discussion by ulissepress** [bcakmakoglu/vue-flow#48] [Ideas]
What about a new feature: at node level we can specify a new property **'resize'** with the following allowed values 'no', 'horizontal', 'vertical', 'both'. 

When the value if not 'no', the user can change the node size using the mouse and related resize handle(s).

See this [discussion](https://github.com/bcakmakoglu/vue-flow/discussions/45) for more details.

**Comment by bcakmakoglu** [comment]
I am not fully convinced of the use-cases as I don't see why the regular nodes would need to be manually resized. 
When the node content changes, the size of a node would usually change "automatically" with a resize observer.

Also a custom node can handle the use-case pretty easily without adding extra options to vue flow.

This post can serve as a sort of interest check is my suggestion, if there's a couple people interested in the feature I will consider it and add it to a future release 😄 

Alternatively I would suggest an "Addon" similar to the [Pathfinding Edge](https://github.com/bcakmakoglu/vue-flow-pathfinding-edge) ([demo here](https://vueflow.dev/docs/addons/pathfinding)).

**Comment by ulissepress** [comment]
ok, got it!

**Comment by bcakmakoglu** [comment]
@ulissepress 
I've thought about this for a while now and will possibly create a plugin with an additional node that can handle resizing, rotating etc.
Meaning it will not be part of the core `@braks/vue-flow` package (to keep it slim) but rather a separate package like `@braks/vue-flow-resizable-node` or something that people can opt into if they need a resizable node.

Would that sound good to your use case? 😄

**Reply by ulissepress** [reply]
Yes absolutely This will be perfect!

**Comment by ulissepress** [comment]
Let me know when this is ready! :-)

**Reply by bcakmakoglu** [reply]
I will try to get to it this weekend but I can't promise it as I got regular work as well 😅
Though it shouldn't be too difficult to create it in a couple hours so you might be lucky 🚀

**Comment by ulissepress** [comment]
Great!

**Reply by bcakmakoglu** [reply]
Will probably take a couple more days.
Progess can be tracked in #136

**Reply by bcakmakoglu** [reply]
It's basically finished but there are some minor bugs that I noticed.
I'll still publish the node package for testing, probably tomorrow at some point.
I can ping you again if the package is published and you can install it and see how it works 😄

**Reply by ulissepress** [reply]
sure, whenever you are ready ping me

**Reply by bcakmakoglu** [reply]
sorry i have been swamped in work but the weahters  going to be bad this weekend so I'll get it done 😅

**Reply by bcakmakoglu** [reply]
@ulissepress 
https://www.npmjs.com/package/@braks/vue-flow-resize-rotate-node

Here you go, this is the initial release. See if it works for!

**Reply by ulissepress** [reply]
The github repo links seems broken (https://github.com/bcakmakoglu/vue-flow/packages/resize-rotate-node)

**Reply by bcakmakoglu** [reply]
No it's just not on the master branch :) Check the develop branch to find the package

**Reply by bcakmakoglu** [reply]
Have you had the chance to try it yet?

**Reply by ulissepress** [reply]
Hello just tried it, ok seems 'nice' but the rotation is now working properly. would be also nice if when i select a node for example dragging it, the resize handles already appears, instead i need another explicit click to have the node resizable.

**Reply by ulissepress** [reply]
another question: where I set the width/height of a node (in the data model I see just x,y). Same for node rotation.

**Reply by ulissepress** [reply]
last question: where i can find the source code of @braks/vue-flow-resize-rotate-node ?

**Reply by bcakmakoglu** [reply]
Ah yeah, I see the rotation issue as well.
I added some fixes for the rotation.

You can apply fixed width/height by either using `node.style.width` or `node.width` and setting it to `300px` for example.
Rotation cannot be pre-set, that's not a built-in function to Vue Flow. 
We'd probably have to add a prop to the ResizeRotateNode for manually setting rotation but then the question is: Why even have a rotatable container if you're gonna set it yourself?
If you just want to rotate an element programatically you can just add a style tag to your custom node and set it to `transform: rotate(30deg)`

Or did I misunderstand you? 😅 

The source code is right here:
https://github.com/bcakmakoglu/vue-flow/tree/master/packages/resize-rotate-node

**Reply by ulissepress** [reply]
My goal is to create a visual editor where the user can create 'diagrams' using different nodes, connecting theme together, etc.

From this point of view, the user should be able to move/resize/rotate a node, and the linked data model should be updated, so when I call toObject() method i can save the state of the diagram (including nodes position, size, rotation).

Hope this is more clear, so what I would like is that when I pass the elements array, for each single node I can also pass the width/height/rotation values (as well as I pass already x and y position).

Hope this clarify
thanks!

**Reply by bcakmakoglu** [reply]
Ah, I understand - I hadn't considered reloading it in the same exact state but of course that makes sense.
I will have to think about the implementation but it should be possible 😄 

Will msg here again as soon as I have an idea on how to do it

**Reply by ulissepress** [reply]
thanks, the best should be to support it natively when you pass the node information (today we pass x and y, tomorrow we (optionally) pass also width, height, rotation). Of course while the user is dragging/resizing all the nodes this data are automatically updated, so when we call toObject() method we get all required data to re-create the graph with the same state.

**Reply by bcakmakoglu** [reply]
Well you *can* actually use `width` and `height` as properties on a node.
Check the [typedocs](https://types.vueflow.dev/interfaces/Node.html) to see all possible properties for a node you can define on initial render.

So that part should actually be copied when calling toObject().
The issue is a) the rotation and b) the fact that the underlying element is not at the same position as the visible node (check your devtools and see how the elements are layered) simply because the visible area needs to be able to move away from the actual position but maybe that can be remedied by aligning them properly after resizing.

Rotation is something that probably won't be built-in into vue flow for the foreseeable future (but who knows) - so rotation values might have to be stored in the `data` object to save & restore it.

**Reply by ulissepress** [reply]
ok clear, supporting natively (i.e. without any specific plugin) width/height properties and interactive resizing will be already a good result!

**Reply by bcakmakoglu** [reply]
@ulissepress there'll be an update of the resize-rotate-node pkg soon; All frame related data (size, rotation etc.) will be saved to `data` object of a resize-rotate node so if you save them you can load them with the correct frame again.

**Reply by ulissepress** [reply]
thank you @bcakmakoglu  to keep me updated on this topic
